src/server/game/Entities/Player/Player.cpp

```diff
@@ -26588,3 +26554,19 @@ bool Player::ValidateAppearance(uint8 race, uint8 class_, uint8 gender, uint8 ha
 
     return true;
 }
+
+void Player::SetRestFlag(RestFlag restFlag)
+{
+    _restFlagMask |= restFlag;
+    SetFlag(PLAYER_FLAGS, PLAYER_FLAGS_RESTING);
+}
+
+void Player::RemoveRestFlag(RestFlag restFlag)
+{
+    _restFlagMask &= ~restFlag;
+    if (!_restFlagMask)
+        RemoveFlag(PLAYER_FLAGS, PLAYER_FLAGS_RESTING);
+
+    if (restFlag == REST_FLAG_IN_TAVERN)
+        InnEnter(time(nullptr), 0);
```

Comment:
When entering a city the areatrigger applies the tavern flag, but it gets removed soon after when actually going further to the city.
At that point this refreshes the inn enter time so the time from entering the area trigger radius and leaving it is lost even though your rested state staid the same.